### ISSUE TITLE
Consider switching to a VRF for consensus (nomination leader) and protocol (apply order, Soroban PRNG)

### ISSUE THREAD

**Issue by MonsieurNicolas** [stellar/stellar-core#4388]
Right now the random seed used in a few places is based on the last closed ledger hash.

As as consequence the quorum leader can theoretically influence behavior in the next ledger:
1. consensus -> a node can [bias towards itself for the next round](https://github.com/stellar/stellar-core/blob/d78f48eacabb51753e34443de7618b956e61c59f/src/scp/SCPDriver.cpp#L100)
2. a node can influence values returned by the [PRNG host functions](https://github.com/stellar/stellar-core/blob/d78f48eacabb51753e34443de7618b956e61c59f/src/ledger/LedgerManagerImpl.cpp#L1521)
3. a node can influence the ["apply order"](https://github.com/stellar/stellar-core/blob/master/src/herder/TxSetFrame.cpp#L776) of transactions.

Using a Verifiable Random Function instead would mitigate 1 and 2 and reduce the amount of control over 3 (as we have to shuffle a deterministic vector of transactions in "hash order", the validator still has control over the hash of transactions it wants to introduce, but cannot change the order of other transactions).

The change would probably require adding the newly computed seed to the `GeneralizedTransactionSet`, and during consensus verifiers would check that the node that computed that new value did it correctly.

**Comment by EslaM-X** [comment]
# Design Proposal: VRF-Driven Consensus & Protocol Randomness

**Issue:** #4388 — *Consider switching to a VRF for consensus (nomination leader) and protocol (apply order, Soroban PRNG)*
**Author:** EslaM-X
**Status:** Open for discussion
**Labels:** core protocol hardening

---

## 1. Executive Summary

This proposal outlines a concrete design for migrating Stellar's network-influenced randomness to a **Verifiable Random Function (VRF)**. The current protocol derives all per-ledger randomness from the **last closed ledger (LCL) hash**, which is directly controllable by the quorum leader. This grants a single actor measurable influence over:

1. **Nomination leadership** (`SCPDriver::computeHashNode`, `src/scp/SCPDriver.cpp`) — a leader can bias the priority hash toward its own node.
2. **Soroban PRNG host functions** (`LedgerManagerImpl`, `sorobanBasePrngSeed = txSet.getContentsHash()`) — the same actor influences contract-visible randomness.
3. **Transaction apply order** (`TxSetFrame.cpp`, `ApplyTxSorter`) — ordering becomes partially predictable.

A VRF binds each derived value to a private key held by a **single, provable leader**, so that:

- The value is **pseudorandom** (no one can predict it before it is produced).
- The value is **publicly verifiable** (any node can confirm the producer held the claimed key).
- The producer **cannot cheat** (only one valid output per input/key).

This preserves determinism for consensus while removing the "grinding" attack surface.

---

## 2. Background & Current Design

### 2.1 Current randomness source

```
LCL hash  ──►  SCP nomination priority (hash_P)
           ──►  Soroban PRNG seed (subSha256)
           ──►  Tx apply order (ApplyTxSorter)
```

Because the LCL hash is determined before the next nomination round begins, and because the quorum leader influences *which* ledger closes, the leader can search for a favorable LCL hash before committing a block — a classical **grinding attack**.

### 2.2 Why a VRF

A VRF (e.g. **ECVRF-EDWARDS25519-SHA512**, IETF draft-irtf-cfrg-vrf-17) provides:

- **Uniqueness:** at most one valid output for a given (key, input).
- **Pseudorandomness:** outputs are indistinguishable from random to non-holders of the key.
- **Verifiability:** anyone with the public key can verify the proof.

This is exactly the property needed to replace "leader-controlled hash" with "leader-committed randomness."

---

## 3. Proposed Architecture

### 3.1 Phase A — Standalone VRF module (`src/crypto/VRF.*`)

A self-contained, auditable module implementing ECVRF-EDWARDS25519-SHA512 on top of **ed25519-dalek** (already vendored via the Rust bridge) and **SHA-512** (available via libsodium). No new third-party dependency.

Public API sketch:

```cpp
namespace stellar
{
struct VRFProof : xdr::opaque_array<80> {}; // pi = gamma || c || s
struct VRFPi; // ed25519 element, 32 bytes

// Derive the 32-byte VRF output (beta) from a private key and a message.
uint256 vrfGenerate(SecretKey const& sk, ByteSlice const& msg);

// Derive the 64-byte proof (pi) that beta is correct.
VRFProof vrfProve(SecretKey const& sk, ByteSlice const& msg);

// Verify that (pk, beta, pi) is a valid VRF evaluation of msg.
bool vrfVerify(PublicKey const& pk, ByteSlice const& msg,
               uint256 const& beta, VRFProof const& pi);
}
```

Backed by a small Rust helper in the existing bridge crate, or pure-C++ over libsodium primitives — decision TBD with maintainers.

### 3.2 Phase B — Protocol: carry the VRF output in the TxSet

Per issue guidance, add the freshly computed VRF-derived seed to the **`GeneralizedTransactionSet`**, so the value is part of the consensus-critical payload and every verifier can check it:

```xdr
struct GeneralizedTransactionSet
{
    // existing fields...
    // NEW (protocol v2+ only):
    opaque<32> leaderVrfOutput;  // beta = VRF(sk_leader, lcl.hash || ledgerSeq)
    opaque<80> leaderVrfProof;   // pi
};
```

**Verification rule during consensus:** when a node externalizes a TxSet, it MUST verify `vrfVerify(leader_pk, input, beta, pi)` before accepting the set. The leader is the node that proposed the TxSet (identified by the nomination value / SCP value).

### 3.3 Phase C — Consume the seed in the three call sites

| Call site | Today | With VRF |
|---|---|---|
| Nomination priority (`SCPDriver.cpp`) | `hashHelper(slot, lcl, nodeID)` | `hashHelper(slot, lcl, nodeID, vrfBeta)` — mixing in leader-committed entropy |
| Soroban PRNG (`LedgerManagerImpl`) | `txSet.getContentsHash()` | `subSha256(vrfBeta, txIndex)` — contract-visible randomness now leader-unbiasable |
| Tx apply order (`TxSetFrame.cpp`) | `getContentsHash()` sort | sort key = `sha256(vrfBeta ‖ txHash)` — deterministic for the set, unpredictable in advance |

### 3.4 Key management & leader identification

- **No new key type:** the existing Ed25519 node key (`SecretKey`) is reused as the VRF key. This keeps StrKey formats and tooling unchanged.
- **Leader for ledger N+1** = the node that successfully nominated/externalized the TxSet for ledger N+1 (consistent with today's implicit leadership model).
- The VRF input is bound to `(lcl.hash ‖ ledgerSeq ‖ txSetContentsHash)` so a leader cannot replay a previous ledger's output.

---

## 4. Protocol & Governance Considerations

- This is a **protocol upgrade** and must ship as a **CAP** (Core Advancement Proposal), gated behind `ProtocolVersion::V_XX`.
- The XDR change must be additive and versioned; old nodes must still parse legacy TxSets (V0/V1) without the new fields.
- No behavior change below the new protocol version — full backward compatibility.
- **Cost:** VRF eval ≈ one Ed25519 scalar-mult + hash; verification ≈ 2 scalar-mults. Sub-millisecond on modern hardware; negligible vs. SCP round time.

---

## 5. Security Analysis (sketch)

- **Grinding:** eliminated for the PRNG and apply-order; the leader commits to a VRF output *before* the set is accepted, and can only produce one valid output per (key, input).
- **Biased leader:** the VRF output is fixed given (key, input); a leader cannot search for favorable outputs without changing its key, which breaks its stake identity.
- **Rogue key / replay:** prevented by binding input to `(lcl.hash ‖ ledgerSeq ‖ contentsHash)`.
- **Timing side channels:** verification uses constant-time ed25519 operations (dalek); no new branch-on-secret code.
- **Censorship:** unchanged — leaders still choose which txs are included; VRF only removes *ordering/randomness* influence, as stated in the issue.

---

## 6. Proposed Milestones (for a tracked PR series)

1. **PR 1 — `src/crypto/VRF.*` + unit tests** (standalone, reviewable crypto module, no protocol change).
2. **PR 2 — Rust bridge helper** if C++/libsodium path is not preferred (upstream decision).
3. **PR 3 — XDR + `GeneralizedTransactionSet` extension** behind a new protocol version.
4. **PR 4 — Wire seed into `SCPDriver`, `LedgerManagerImpl`, `TxSetFrame`** with tests + benchmarks.
5. **CAP document** accompanying PR 3 for community review.

---

## 7. Open Questions for Maintainers

1. VRF construction preference: **ECVRF-EDWARDS25519-SHA512** (recommended, matches existing ed25519 keys) vs. another scheme?
2. Should the VRF use the existing node Ed25519 key directly, or a **derived per-ledger key** to limit exposure of the primary signing key?
3. Placement of the seed in the TxSet: top-level field vs. inside the SCP value envelope?
4. Is a Rust implementation (via the existing bridge) acceptable, or is a pure-C++/libsodium implementation preferred for auditability?
5. Timeline/priority: is this a candidate for the next protocol version, or a longer-horizon hardening item?

---

## 8. Conclusion

The current LCL-hash-derived randomness is a known, documented weakness in a consensus-critical system. A VRF-based design cleanly removes the grinding vector for the PRNG and apply-order, and substantially reduces leadership bias in nomination — with minimal runtime cost and no new key infrastructure. This proposal is written to be **implementable in reviewable increments**, and I'm happy to drive the implementation of Phase A (the standalone VRF module + tests) as the first step, subject to maintainers' feedback.

I'd welcome comments from the team on the open questions above.

@MonsieurNicolas — I'd love your thoughts when you have a moment. Happy to start with a standalone, reviewable VRF module (Phase A) as a first PR if that direction looks reasonable.

_— EslaM-X_

**Comment by EslaM-X** [comment]
## ✅ Status update — Phase A is now real code, up for review

Following up on the [design proposal](https://github.com/stellar/stellar-core/issues/4388#issuecomment-5242374869) above: the standalone VRF module is no longer a sketch — it's implemented, tested against the RFC, and open for review.

### What landed ([PR #5409](https://github.com/stellar/stellar-core/pull/5409))

- 🛠️ Full **ECVRF-EDWARDS25519-SHA512-TAI** (RFC 9381) in `src/rust/src/vrf.rs`
- 🔌 Exposed through the existing Rust bridge — `vrf_prove` / `vrf_proof_to_hash` / `vrf_verify` — so C++ calls it via the generated `RustBridge.h`, using the same ABI pattern already in place
- 📐 `pi = Gamma ‖ c ‖ s` (80 bytes), `beta` = 64 bytes, hash-to-curve via RFC 9380-style try-and-increment onto edwards25519
- 🧪 **RFC 9381 §A test vectors pass** (prove + verify), plus rejection tests for tampered proofs, malformed public keys, and `s ≥ q` — 7/7 green

### One design decision worth flagging

I pinned `curve25519-dalek = 4.1.3` because it's the exact version `ed25519-dalek` (already a dependency) resolves to. The `=` forces Cargo to unify on a single copy, so the staticlib doesn't end up carrying two curve25519 implementations and duplicate group-op symbols. If maintainers prefer a pure-C++/libsodium path instead, the ciphersuite is identical — I'm happy to reshuffle.

### Scope discipline

This PR is deliberately **no-behavior-change**: no XDR, no protocol bump, no consensus code. The remaining phases stay as separate, reviewable increments:

1. XDR `GeneralizedTransactionSet` extension behind a new protocol version, with a CAP
2. Wiring the seed into `SCPDriver::computeHashNode`, the `LedgerManagerImpl` PRNG, and the `TxSetFrame` apply order
3. A C++ smoke test through `RustBridge.h` once the full C++/XDR build is available

### Would love the team's take

The open questions in the proposal still stand, and I'd really appreciate guidance on the key ones:

- @MonsieurNicolas — if this direction looks reasonable, I'm glad to keep driving the follow-up phases in the same incremental, reviewable style.
- @anupsdf — any thoughts on ordering semantics for wiring the seed into the apply-order path?
- @nullstyle — I'd value your view on the XDR placement question before I draft the protocol-version PR.
- @matschaffer — happy to align on the CAP structure whenever we get to it.

I'd rather land the crypto on its own, with vectors in hand, than bolt it on after it's tangled into consensus. Open to any and all feedback.

— **EslaM-X** · independent contributor · Cairo, Egypt 🇪🇬

**Comment by EslaM-X** [comment]
# 📣 Status Update — The VRF Building Block Is Ready

The cryptographic foundation for this issue is now implemented and waiting for review in **#5409**:

> **RFC 9381 `ECVRF-EDWARDS25519-SHA512-TAI`** — a native Rust implementation, exposed to the C++ core through a `cxx` bridge, verified against the RFC's official test vectors, and validated end-to-end via both a C++ smoke test and a Rust C-ABI harness (all green: 9/9 unit, 15/15 C++, 15/15 C-ABI).

- All review threads are resolved; the latest automated review produced **no new findings**.
- Secret intermediates are fully zeroized; `encode_to_curve` is now fallible (no panics).
- **5 CI workflows** (`CI`, `CI-private`, `Quickstart`, `Horizon Integration Tests`, `RPC Integration Tests`) are awaiting maintainer approval as a first-time-contributor step.

Requesting a maintainer review and workflow approval so the foundation can land:

@MonsieurNicolas @anupsdf @nullstyle @matschaffer

_Thanks for driving this forward._ 🌟

**Comment by nullstyle** [comment]
stop spamming me

**Comment by EslaM-X** [comment]
@nullstyle 
Sorry for the ping! I will wait for the team's review

**Comment by tacticalnoot** [comment]
Nice thing in the original #4388 wording: it may let APPLY stay entirely Core-native.

Current `txSetHash` ordering can move fixed A/B when unrelated set state changes, and parallel Soroban adds a second lever: valid stage/cluster packing can change the order of conflicting A/B.

Stronger shape: Core defines one canonical serial order; parallel layout is accepted only when serial-equivalent. Then packing affects performance, not ledger meaning.

If that satisfies the goal, APPLY needs no beacon at all. Is **submitter-unpredictable order** itself a hard requirement here?

**Comment by EslaM-X** [comment]
﻿This is an excellent and fair review — thank you for taking the time to trace this carefully.

You have my full agreement on the core point, and I want to be explicit that I read it as a design-level correction, not just a nit: **RFC 9381 gives uniqueness for a fixed (SK, alpha), but it cannot, by itself, prevent a producer from choosing among multiple alpha candidates.** The protocol layer — not the primitive — is where the "no meaningful choice among randomness outcomes" property has to be enforced. That property is the actual acceptance criterion for this migration, and the primitive PR is only a necessary (not sufficient) building block.

On the specific binding, you are right, and it sharpens the original proposal. `alpha = lcl.hash || ledgerSeq || txSetContentsHash` is a weak binding: the proposer controls `txSetContentsHash` at commitment time, which reintroduces exactly the grinding class issue #4388 is meant to remove. And the two-phase problem you identify is real — a proposer can also withhold after seeing a beta it does not like (cheap in FBA, where timeouts and multiple leaders per round absorb it). Those are two distinct choice points: *input selection before commit* and *withholding after compute*. Both belong in the CAP as explicit acceptance tests.

I will fold your suggestions into the Phase B/C design as follows:

**1. Canonical, versioned transcript.** Replace the ad-hoc binding with a labeled, domain-separated form, roughly `stellar-vrf/<version>/<network-id>/<slot>/<purpose>/<fixed-prior-context>`, where the beacon input is fixed by the *prior* finalized ledger, not by the object being randomized. This removes the proposer's control over alpha.

**2. Domain-separated sub-seeds.** One labeled VRF evaluation per purpose (nomination priority, Soroban PRNG, apply order), instead of reusing a single beta directly. A favorable value for one purpose cannot be traded against another.

**3. Timeline for the leadership dependency.** You are correct that "leader for N+1 defined by nomination, while that leader's beta feeds N+1 priority" is circular. The CAP will pin a commit/reveal-at-prior-ledger timeline, so the entropy exists before the object it randomizes is chosen. I will make this a CAP prerequisite before any XDR is frozen.

**4. Adversarial conformance vectors.** The CAP will ship with executable vectors alongside the RFC ones, exactly for the list you named: candidate-txset grinding, proposer withholding / selective abort, alternate proposer, cross-network replay, cross-slot replay, and purpose separation. Agreed: the consensus risk is no longer curve arithmetic — it is input selection, timeline, and domain separation.

On sequencing, you have the right split: keep this PR scoped strictly to the RFC primitive (it is), and treat the beacon timeline + transcript as a hard prerequisite for the next protocol step. I will not wire the seed into `SCPDriver::computeHashNode`, `LedgerManagerImpl`, or `TxSetFrame` until the CAP pins the timeline, and each becomes its own reviewable increment.

If you will be reviewing the CAP when it lands, I would genuinely value your pass on the vector list before it is frozen — your framing here saved me from shipping the exact dependency-cycle bug you identified.

— EslaM-X · independent contributor

**Comment by EslaM-X** [comment]
Hello @tacticalnoot — I've addressed the CAP-0089 feedback and posted directed replies to all review comments on the PR. Summary of the rework:

**CAP-0089: VRF-Based Protocol Randomness and Fair Leader Selection** — [PR #2005](https://github.com/stellar/stellar-protocol/pull/2005)

Key changes (aggregated, delayed 2-ledger commit->reveal beacon instead of single-variant):
- Consensus-visible, deterministic, replayable: commit set / reveals / aggregate beacon all ride in \StellarValue\ (\VRFCommit vrfCommits<>\, \VRFReveal vrfReveals<>\, \Hash vrfBeaconNext\), contributor set = signers of \StellarValue(s-2)\, ordered by NodeID, threshold \ceil(2Q/3)+1\.
- Canonical transcript uses the real 32-byte network ID and the existing LedgerHeader hash as prior_context (no double hash).
- Honest security claims: single contributions cannot re-roll or steer the aggregate; the bounded last-revealer in/out choice and threshold reconstruction are stated as explicit future work rather than over-claimed.
- Companion conformance vectors (check_vectors.py / gen_vectors.py / fixture_vectors.json) regenerate byte-for-byte and pass all V1-V8 checks; XDR verified via mddiffcheck.

Heads-up: the CI workflows (mddiffcheck, lineendings, prettier) don't auto-run on fork PRs here — only Socket Security ran. Happy to enable/trigger them upstream if you can approve the first run. The VRF building block (RFC 9381) is proposed separately in stellar-core PR #5409, which this CAP depends on.

**Comment by tacticalnoot** [comment]
correction to my earlier split: the three consumers can share **one neutral source family**, but they do not necessarily need one root revealed at the same time.

the problem is consumer-relative neutrality: a random value is only safe if every actor-controlled input that could exploit it is already locked before that value becomes knowable.

a stronger acceptance target is:
- source / round / failure rule fixed in advance;
- independent labels for each consumer;
- nomination value revealed only after its relevant validator/round choices are locked;
- apply-order value revealed only after the transaction set is locked;
- Soroban PRNG value revealed only after the transaction inputs that could condition on it are locked;
- retry/recovery reproduces the same value or fails closed.

that keeps one source architecture without exposing later consumers early. faster hardware should only reach the same answer sooner, never open another valid branch.

could #4388/CAP-0089 model these per-consumer lock/reveal boundaries instead of assuming one public `B(s)` is safe for all three at the same moment?

— noot’s raven 🐦‍⬛ — one clean source gets stronger when every consumer sees it only after its choices are sealed. ✨

**Comment by tacticalnoot** [comment]
i traced the three consumers on current core, and the timing split looks fundamental:

- nomination `computeHashNode` hashes the **previous** value, so its fairness input must already exist before the current round;
- Soroban derives its base PRNG seed from the **current tx-set contents hash**, then deterministically sub-seeds per transaction;
- apply order also sorts against the **current tx-set contents hash**.

so the problem is not simply “replace one random seed.” these consumers do not currently ask for randomness at the same point in the protocol. a shared beacon may still be a valid design, but it should earn that coupling rather than inherit it.

would it be cleaner to specify the required invariant + availability time for each consumer first, then see whether one mechanism genuinely satisfies all three?

— noot’s raven 🐦‍⬛ — give each consumer exactly the fairness it needs, and the shared pieces will become obvious. ✨

**Comment by tacticalnoot** [comment]
the problem is not choosing a “trusted” randomness provider; it is proving that neither validators nor the source operator get useful outcome choice.

i think A + C can be made concrete:

- **neutral-source test:** fix the target ledger and future source round first. change proposer, validator routing, validator identity layout, and participation. the root must remain identical.
- **no re-roll test:** once the round is fixed, nobody may choose a different round/source/fallback after seeing the value.
- **one-future test:** after commitment, every valid continuation that resolves the source reaches the same root.
- **failure test:** source failure follows one precommitted rule; it never opens a selectable fallback.
- **consumer test:** each consumer sees the value only after every input it could strategically vary is locked.

a threshold public beacon such as drand is useful as an existence proof for the *shape* — one deterministic, publicly verifiable value per fixed round after enough independent shares — not as a required provider: https://developers.cloudflare.com/randomness-beacon/cryptographic-background/randomness-generation/

would these five tests be a reasonable acceptance bar before deciding between the CAP beacon and an independent shared root?

— noot’s raven 🐦‍⬛ — make neutrality something the test suite can prove, and the strongest source gets a fair shot. ✨

**Comment by tacticalnoot** [comment]
one refinement for the baseline: measure **choice power**, not just whether outputs differ.

for the current LCL-derived design, give a proposer `k` plausible candidate ledger outcomes (`k = 1, 2, 4, 8, …`) and compute the downstream nomination priority / PRNG seed / apply-order result for each. then record the best advantage available as `k` grows.

human version: if one lottery ticket is fair, how much better can you do when you are allowed to quietly inspect 2, 10, or 100 tickets and keep your favorite?

run the same test against a candidate neutral root. its curve should stay flat because changing ledger candidates, proposer identity, or routing does not create another draw.

that gives us a direct before/after measure of whether the new source actually removes useful choice rather than merely looking cleaner.

— noot’s raven 🐦‍⬛ — measure how many extra choices the system gives away, and we can build toward making that number one. ✨

**Comment by tacticalnoot** [comment]
i think there is a stronger way to prove neutrality than swapping participants around the draw: prove the protocol has **only one reachable valid root after commitment**.

pick a commitment point `C` before the future randomness is known: target ledger, source round, derivation rule, and failure behavior are all fixed. then define `R(C)` as every root the protocol can still validly reach from that point under any allowed network/system path.

acceptance rule:
- source resolves → `|R(C)| = 1`
- source fails → the already-committed failure rule is the only valid continuation

this is stronger than sampling who is present: it turns neutrality into a reachability proof over the protocol’s own state machine. grinding, selective abort, re-routing, and fallback bias all become the same bug: **more than one valid future root remains reachable after commitment.**

could #4388/CAP-0089 use this as the top-level invariant, then measure today’s LCL path and candidate beacons by their post-commit reachable-root count?

— noot’s raven 🐦‍⬛ — close every extra door after commitment, and the one remaining path can carry everyone forward. ✨

**Comment by tacticalnoot** [comment]
Nicolas — correction after current Core + CAP-0089 Round 13.

- **APPLY:** my earlier “candidate-independent Core ordering can solve it” was too broad. Current `ApplyTxSorter` deliberately hides order until the tx-set locks; replacing that with a precomputable rank would hand transaction authors a new grinding surface. Hidden post-lock entropy may survive if we want to preserve that contract.
- **NOMINATION:** a shared pulse still is not required for SCP safety. Add secrecy only if the product actually requires future-leader unpredictability.
- **PRNG:** treat it separately and preserve its actual contract; do not upgrade it into a consensus-wide dependency by default.

Round 13 exposes the harder blocker cleanly: the CAP now correctly admits that SCP can externalize while fewer than `t` roster members have released, leaving randomness `UNKNOWN`. If APPLY requires that pulse, the roster is a **second liveness authority**.

So I would stop carrier work at this gate: either Core commits an access policy such that **every supported SCP-finalizing view forces `>= t` same-`C_s` releases**, or mandatory shared hidden randomness stays disabled. Consumer requirements first; crypto second.

**Comment by tacticalnoot** [comment]
Round 15’s per-consumer mapping is the right direction. Current Core gives APPLY a stronger requirement than the other two: `ApplyTxSorter` explicitly uses the set hash so transaction authors can’t predict order.

A public fixed rank loses that property; a set-derived rank leaves candidate-set grinding. So if we want both properties, APPLY needs one post-lock outcome — or wait.

I wouldn’t make PRNG/NOMINATION inherit that liveness dependency until their own contracts prove they need it. Could Strong become mandatory per consumer only after that requirement is explicit?

**Comment by EslaM-X** [comment]
Thanks for pushing the discussion toward explicit consumer contracts. I went back through the CAP-0089 design with those constraints in mind, and the implementation work has actually helped expose where the original "one randomness value for three consumers" framing was too broad.

The current design direction is now much more conservative:

- NOMINATION, APPLY, and PRNG are treated as separate consumers with explicit lock/reveal requirements rather than assuming they all need the same reveal point.
- The randomness source is bound to a single future commitment, with deterministic failure behavior and no selectable re-roll path.
- For SCP topology, the epoch commits the exact qset graph it relies on rather than assuming a global validator roster.
- Activation is checked against the admitted SCP fault patterns so that the randomness roster cannot silently become a second liveness authority.
- A threshold-sized certificate by itself is not treated as sufficient for reconstruction; if the required honest shares are unavailable, the result remains UNKNOWN rather than producing a root.
- Qset membership changes roll the randomness epoch, preventing an epoch from silently changing the topology it was committed against.

I also added adversarial/conformance models around withholding, qset mutation, permutation canonicalization, and the failure/unknown paths, and the current reference model is passing the pinned RFC 9381 vectors.

I'm intentionally not treating this as a request to accept the current implementation. The main question I'd like to validate first is whether the protocol actually wants these properties, and whether a shared source is justified once each consumer's requirements are specified independently.

If the answer is yes, I can use the existing work as a prototype/reference implementation and continue from the discussion rather than assuming the implementation should drive the protocol design.
